sonic_installer/bootloader/grub.py

```diff
@@ -81,6 +82,34 @@ def remove_image(self, image):
         run_command('grub-set-default --boot-directory=' + HOST_PATH + ' 0')
         click.echo('Image removed')
 
+    def verify_binary_image(self, image_path):
+        if not os.path.isfile(image_path):
+            return False
+
+        # Get running platform's ASIC
+        sonic_version_yml = open('/etc/sonic/sonic_version.yml', 'r')
+        asic_type = re.search(r"asic_type: (\S+)", sonic_version_yml.read()).group(1)
+        sonic_version_yml.close()
+        # click.echo('Get running platform ASIC... %s' % asic_type)
+
+        # Get installing image's ASIC
+        p1 = subprocess.Popen(["cat", "-v", image_path], stdout=subprocess.PIPE, preexec_fn=default_sigpipe)
```

Comment:
>p1 [](http://example.com/codeflow?start=8&length=2)

Did you handle the return code of each process?

Reply:
https://github.com/Azure/sonic-utilities/blob/master/sonic_installer/bootloader/onie.py#L33
I am referring from here. 
If 'image_asic' is missing in binary image, final result string will be empty. It will not impact the comparison.
p1 will throw exception if image_path doesn't exist, but this is checked in previous step.
Is this solve your question?

Reply:
I see the existing code sample. My question is about general usage of `subprocess.Popen` and pipe. If upstream process exit with a non-zero code, does the whole pipe stop immediately? will the last process `p3` get an non-zero error code? or do we need to check all the return code of p1, p2, p3?

Reply:
Checked locally. Downstream could continue to run even upstream exit non-zero code and whole pop will not stop.
I think it is fine.

Reply:
If you does not check non-zero return code in the pipeline, you will get an empty `image_asic`, and this function will return True. This is very dangerous. Instead, we need to report this as an unexpected failure, or at least return False to be safe.

Reply:
Agreed. Change to return False.